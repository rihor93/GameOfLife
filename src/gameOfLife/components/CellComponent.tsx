import { PropsWithChildren, useState } from "react"
import './CellComponent.css'

/**
     * интерфейс ячейки сетки, содержит id и жизненный статус
     */
export interface Cell {
    id: number,
    status: CellStatus,
}

/**
 * жизненные статусы ячейки
 */
export enum CellStatus{
    Alive,
    AliveOld,
    Dead,
}

export const CellComponent: React.FC<
PropsWithChildren<{ value: Cell}>
>= ({value}) => {

    const [cellState, setCellState] = useState(value);


    /**
     * функция меняет состояние ячейки
     */
    function updateCell(): void{

        setCellState((previewState) => ({id: previewState.id, status: previewState.status == CellStatus.Alive ? CellStatus.Dead : CellStatus.Alive}));
    }

    return (
    <div data-testid="cellcomponent" className={cellState.status === CellStatus.Alive ? "cell alive" : (cellState.status === CellStatus.AliveOld ? "cell alive old" : "cell dead")} onClick={updateCell}>
        {cellState.id}
    </div>)
}