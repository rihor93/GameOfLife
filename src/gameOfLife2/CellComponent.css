@keyframes alive_old_animation {
    from {
        background-color: red;
    }

    to {
        background-color: yellow;
    }
}

@keyframes alive_dead_animation {
    from {
        background-color: red;
    }

    to {
        background-color: black;
    }
}

@keyframes old_alive_animation {
    from {
        background-color: yellow;
    }

    to {
        background-color: red;
    }
}

@keyframes old_dead_animation {
    from {
        background-color: yellow;
    }

    to {
        background-color: black;
    }
}

@keyframes dead_alive_animation {
    from {
        background-color: black;
    }

    to {
        background-color: red;
    }
}

@keyframes dead_old_animation {
    from {
        background-color: black;
    }

    to {
        background-color: yellow;
    }
}