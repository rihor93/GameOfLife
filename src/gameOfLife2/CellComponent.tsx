import React, { useEffect, useState } from "react";


export interface Cell {
    id: number,
    status: CellStatus,
}

/**
 * жизненные статусы ячейки
 */
export enum CellStatus {
    Alive,
    AliveOld,
    Dead,
}

export type Props = {
    id: number,
    status: CellStatus,
    onClick: (id: number) => void,
    className?: string,
};

const CellComponent: React.FC<Props> = (props) => {
    //console.log('CellComponent')
    const [id, setId] = useState(props.id);
    const [status, setStatus] = useState(props.status);

    useEffect(() => {
        setStatus(props.status);
    }, [props.status])


    return (<div className={props.className} data-testid="cellcomponent" /*className={status === CellStatus.Alive ? "cell alive" : (status === CellStatus.AliveOld ? "cell old" : "cell dead")}*/
        onClick={() => props.onClick(id)}>
        {id}
    </div>);
}



export default CellComponent;

