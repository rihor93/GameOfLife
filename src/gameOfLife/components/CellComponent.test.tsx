import { render, screen } from '@testing-library/react';
import { CellComponent, CellStatus } from './CellComponent';
import renderer from 'react-test-renderer';

describe('CellComponent tests', () => {
    it('show id on screen', () => {
        let cellData = { id: 1, status: CellStatus.Alive };
        render(<CellComponent value={cellData} />);
        const allElements = screen.getAllByText('1');

        expect(allElements).toHaveLength(1);
    });

    it('show alive element with class alive', () => {
        let cellData = { id: 1, status: CellStatus.Alive };
        const { container } = render(<CellComponent value={cellData} />);
        const allElements = container.getElementsByClassName("cell alive")

        expect(allElements).toHaveLength(1);
    });

    it('show alive element with class dead', () => {
        let cellData = { id: 1, status: CellStatus.Alive };
        const { container } = render(<CellComponent value={cellData} />);
        const allElements = container.getElementsByClassName("cell dead")

        expect(allElements).toHaveLength(0);
    });

    it('show dead element with class dead', () => {
        let cellData = { id: 1, status: CellStatus.Dead };
        const { container } = render(<CellComponent value={cellData} />);
        const allElements = container.getElementsByClassName("cell dead")

        expect(allElements).toHaveLength(1);
    });

    it('show aliveold element with class aliveold', () => {
        let cellData = { id: 1, status: CellStatus.AliveOld };
        const { container } = render(<CellComponent value={cellData} />);
        const allElements = container.getElementsByClassName("cell old")

        expect(allElements).toHaveLength(1);
    });

});

describe('CellComponent Snapshottest', () => {
    it('Alive cell snapshot', () => {
        let cellData = { id: 1, status: CellStatus.Alive };
        const cellContainer = renderer
            .create(<CellComponent value={cellData} />)
            .toJSON();
        expect(cellContainer).toMatchSnapshot();
    });

    it('Alive old cell snapshot', () => {
        let cellData = { id: 12, status: CellStatus.AliveOld };
        const cellContainer = renderer
            .create(<CellComponent value={cellData} />)
            .toJSON();
        expect(cellContainer).toMatchSnapshot();
    });

    it('Dead cell snapshot', () => {
        let cellData = { id: 47, status: CellStatus.Dead };
        const cellContainer = renderer
            .create(<CellComponent value={cellData} />)
            .toJSON();
        expect(cellContainer).toMatchSnapshot();
    });

});