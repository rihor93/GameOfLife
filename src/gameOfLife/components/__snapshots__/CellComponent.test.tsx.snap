// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellComponent Snapshottest Alive cell snapshot 1`] = `
<div
  className="cell alive"
  onClick={[Function]}
>
  1
</div>
`;

exports[`CellComponent Snapshottest Alive old cell snapshot 1`] = `
<div
  className="cell alive old"
  onClick={[Function]}
>
  12
</div>
`;

exports[`CellComponent Snapshottest Dead cell snapshot 1`] = `
<div
  className="cell dead"
  onClick={[Function]}
>
  47
</div>
`;
